### PR TITLE
Added methods for nilpotent, abelian and cyclic numbers.

### PR DESCRIPTION
**Pull request by OrestisVaggelis**
#### Brief description of what is fixed or changed
Implemented methods to check if a number is nilpotent, abelian or cyclic. Details about the algorithm can be found in
[this](http://www2.math.ou.edu/~shankar/papers/nil2.pdf) paper.

#### Release Notes

* combinatorics
  * Added methods for nilpotent, abelian and cyclic numbers.

### CODE REVIEW

File: AUTHORS [hunk 1]
```diff
@@ -1133,3 +1133,13 @@ Tom Fryers <61272761+TomFryers@users.noreply.github.com>
...
+Siddhant Jain <siddhantashoknagar@gmail.com>
+Tirthankar Mazumder <63574588+wermos@users.noreply.github.com>
+Sumit Kumar <mr.sumitkrr@gmail.com>
+Shivam Sagar <technoshivam12@gmail.com>
+Gaurav Jain <gjain369@gmail.com>
+Andrii Oriekhov <andriyorehov@gmail.com>
+Luis Talavera <luisfertalavera15@gmail.com>
+Arie Bovenberg <a.c.bovenberg@gmail.com>
+Carson McManus <carson.mcmanus1@gmail.com>
+Jack Schmidt <1107865+jackschmidt@users.noreply.github.com>
```
Comment (oscarbenjamin) [inline AUTHORS:1145]: The changes to this file should be reverted. The AUTHORS file is updated at the time of the next release. It's only necessary to add your details in the .mailmap file.

File: .mailmap [hunk 2]
```diff
@@ -831,7 +831,7 @@ Ondřej Čertík <ondrej@certik.cz> <ondrej.certik@gmail.com>
 Ondřej Čertík <ondrej@certik.cz> <ondrej@certik.us>
 Ondřej Čertík <ondrej@certik.cz> ondrej.certik <devnull@localhost>
 Or Dvory <gidesa@gmail.com>
-OrestisVaggelis <orestis22000@gmail.com>
+OrestisVaggelis <orestis22000@gmail.com> Orestis Vaggelis <orestisvaggelis@mail.com>
```
Comment (oscarbenjamin) [inline .mailmap:834]: Which name and email do you want to appear in the AUTHORS file?

Comment (OrestisVaggelis) [reply .mailmap:834]: I'd prefer Orestis Vaggelis <orestisvaggelis@mail.com> to appear, but without the first one, I was getting an error during the tests.

File: sympy/combinatorics/group_numbers.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    >>> is_nilpotent_number(21)
+    False
+    >>> is_nilpotent_number(randprime(1, 30)**12)
+    True
+
+    References
+    ==========
+
+    .. [1] Pakianathan, J., Shankar, K., Nilpotent Numbers,
+           The American Mathematical Monthly
```
Comment (jksuom) [inline sympy/combinatorics/group_numbers.py:25]: There should be more details like year, issue, pages..

File: .mailmap (hunk 2 above)
Comment (oscarbenjamin) [reply .mailmap:834]: Then the above should be reversed like:
```
Orestis Vaggelis <orestisvaggelis@mail.com>  OrestisVaggelis <orestis22000@gmail.com>
```
Also you should remove the line `OrestisVaggelis <orestis22000@gmail.com>` from the AUTHORS file. The new name will then be added to AUTHORS automatically at the time of the next release.

File: AUTHORS [hunk 4]
```diff
@@ -1133,3 +1132,9 @@ Tom Fryers <61272761+TomFryers@users.noreply.github.com>
 Zouhair <zouhair.mahboubi@gmail.com>
 zzj <29055749+zjzh@users.noreply.github.com>
 shubhayu09 <guptashubhayu601@gmail.com>
+Siddhant Jain <siddhantashoknagar@gmail.com>
+Tirthankar Mazumder <63574588+wermos@users.noreply.github.com>
+Sumit Kumar <mr.sumitkrr@gmail.com>
+Shivam Sagar <technoshivam12@gmail.com>
+Gaurav Jain <gjain369@gmail.com>
+Andrii Oriekhov <andriyorehov@gmail.com>
```
Comment (oscarbenjamin) [inline AUTHORS:1140]: These lines should also be removed.

**Comment by OrestisVaggelis** [comment]
Thank you for your feedback. I've now reverted the changes made in the AUTHORS file, updated the .mailmap file and added more information about the reference.

File: AUTHORS [hunk 5]
```diff
@@ -4,7 +4,7 @@ those who explicitly didn't want to be mentioned. People with a * next
 to their names are not found in the metadata of the git history. This
 file is generated automatically by running `./bin/authors_update.py`.
 
-There are a total of 1127 authors.
+There are a total of 1137 authors.
```
Comment (oscarbenjamin) [inline AUTHORS:7]: This change also needs to be reverted.

When looking at the PR you can click "files changed" at the top to see what is actually being changed.

Comment (OrestisVaggelis) [reply AUTHORS:7]: So there should be 1126 authors now, since I removed myself, right?

Comment (oscarbenjamin) [reply AUTHORS:7]: The number will be updated at the time of the next release. For now it is best just to leave it unchanged.

The reason for not changing things like this is that if any two PRs changes the same line of the same file then there will be merge conflicts. If every PR tries to change this line then every PR will have merge conflicts.

**Comment by oscargus** [comment]
Should this be Abelian group in the docs? Wolfram seems to use Abelian, while Wikipedia and many other sources use abelian. Since Abel is a person it is not obvious to me as non-native.

**Comment by OrestisVaggelis** [comment]
> Should this be Abelian group in the docs? Wolfram seems to use Abelian, while Wikipedia and many other sources use abelian. Since Abel is a person it is not obvious to me as non-native.

Other docs in sympy seem to use abelian as well (see permutation groups for example). Also the paper referenced uses a lower case a too.

**Pull request merged**